File: applications/spring-geode-showcase/src/test/java/com/vmware/spring/geode/showcase/controller/AccountControllerTest.java
package com.vmware.spring.geode.showcase.controller;

import com.vmware.spring.geode.showcase.domain.Account;
import com.vmware.spring.geode.showcase.repostories.AccountRepository;
import nyla.solutions.core.patterns.creational.generator.JavaBeanGeneratorCreator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class AccountControllerTest
{
    @Mock
    private AccountRepository accountRepository;

    private AccountController subject;
    private Account account;

    @BeforeEach
    void setUp()
    {
        account = JavaBeanGeneratorCreator.of(Account.class).create();
        subject = new AccountController(accountRepository);


    }

    @Test
    @DisplayName("Given account When save Then Can get account by Id")
    void createRead()
    {
        when(accountRepository.findById(account.getId())).thenReturn(Optional.of(account));
        subject.save(account);
        verify(accountRepository).save(account);
        assertEquals(account,subject.findById(account.getId()).get());

    }

    @Test
    @DisplayName("Given saved account When delete Then repository delete called")
    void delete()
    {
        subject.deleteById(account.getId());
        verify(accountRepository).deleteById(account.getId());
    }
}

File: applications/account-rest-service/src/test/kotlin/com/vmware/financial/open/banking/account/controller/AccountControllerTest.kt
package com.vmware.financial.open.banking.account.controller

import com.vmware.financial.open.banking.account.domain.Account
import com.vmware.financial.open.banking.account.service.AccountService
import nyla.solutions.core.patterns.creational.generator.JavaBeanGeneratorCreator
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import java.util.*

internal class AccountControllerTest {

    private lateinit var account: Account

    private lateinit var accountService: AccountService

    private lateinit var subject : AccountController

    @BeforeEach
    internal fun setUp() {
        account = JavaBeanGeneratorCreator.of(Account::class.java).create()
        accountService = mock<AccountService>(){
            onGeneric { createAccount(any())} doReturn account
            onGeneric { findAccountById(anyString(),anyString())} doReturn Optional.of(account)
        }
        subject = AccountController(accountService)
    }

    @Test
    fun create() {
        val bankId = "expected"
        var response = subject.createAccount(bankId,account)
        verify(accountService).createAccount(account)
    }

    @Test
    internal fun getAccountById_foundAccount() {
        val bankId = "expected"
        var actualHttpEntity : ResponseEntity<Account> = subject.getAccountById(bankId,account.id)
        assertEquals(actualHttpEntity.statusCode ,HttpStatus.OK)
        assertEquals(account,actualHttpEntity.body);
    }

    @Test
    internal fun getAccountById_notFoundAccount() {

        accountService = mock<AccountService>(){
            onGeneric { findAccountById(anyString(),anyString())} doReturn Optional.empty()        }
        subject = AccountController(accountService)

        val bankId = "expected"
        var actualHttpEntity : ResponseEntity<Account> = subject.getAccountById(bankId,account.id)
        assertEquals(actualHttpEntity.statusCode ,HttpStatus.NOT_FOUND)
    }

}